fix(india): duplicate qrcode and hide button

maharshivpatel opened the pull request:
Problem: allows qrcode generation even if it exists and makes API call even though data will most likely be available in signed_qr_code field.

Solution: I have added logic on the server & client side to prevent the duplicate generation of qrcode and use signed_qr_code if available this will allow generating qrcode even after 2 days. if all fails it will get qrcode using api.

Note: I have added __unsaved check on the client side this is for future proofing.

> currently, there is no way to change anything in the invoice once you generate IRN however this is not the correct approach as there are many fields that have nothing to do with Invoice details that we send to the IRP ( government ) and we should be able to change fields that don't affect data we sent to the IRP. I am working on it and will send separate PR or this.

maharshivpatel committed:
fix(india): duplicate qrcode and hide button

problem: allows qrcode generation even if it exists and makes api call even though data will most likely be available in signed_qr_code field.
solution: i have added logic on server & client side to prevent duplicate generation of qrcode and use signed_qr_code if available this will allow to generate qrcode even after 2 days. if all fails it will get qrcode using irn api.

maharshivpatel committed:
sider fix

indentation and semi colon fix.

maharshivpatel committed:
fix sider

indentation fix again

nextchamp-saqib commented on erpnext/regional/india/e_invoice/utils.py:
```diff
@@ -1010,13 +1010,29 @@ def bulk_generate_irn(invoices):
-			frappe.msgprint(_("QR Code attached to the invoice"), alert=True)
-		else:
-			frappe.msgprint(_("QR Code not found for the IRN"), alert=True)
+		is_qrcode_file_attached = frappe.db.exists(
```
```suggestion
		is_qrcode_file_attached = self.invoice.qrcode_image and frappe.db.exists(
```

nextchamp-saqib commented on erpnext/regional/india/e_invoice/utils.py:
```diff
@@ -1010,13 +1010,29 @@ def bulk_generate_irn(invoices):
+					frappe.msgprint(_("QR Code attached to the invoice."), alert=True)
+				else:
+					frappe.msgprint(_("QR Code not found for the IRN"), alert=True)
+		elif is_qrcode_file_attached and self.invoice.qrcode_image:
```
```suggestion
		else:
```

nextchamp-saqib commented on erpnext/regional/india/e_invoice/einvoice.js:
```diff
@@ -175,27 +175,44 @@ erpnext.setup_einvoice_actions = (doctype) => {
+						}
+					}
+				}
+				if (!is_qrcode_attached || !qrcode_image) {
```
```suggestion
				if (!is_qrcode_attached) {
```

If `qrcode_image` is not set then `is_qrcode_attached` will always be `false`

nextchamp-saqib commented on erpnext/regional/india/e_invoice/utils.py:
```diff
@@ -1010,13 +1010,29 @@ def bulk_generate_irn(invoices):
+		)
+		if not is_qrcode_file_attached:
+			if self.invoice.signed_qr_code:
+				self.attach_qrcode_image(update_url=True)
```
```suggestion
				self.attach_qrcode_image()
				frappe.db.set_value("Sales Invoice", self.invoice.name, "qrcode_image", self.invoice.qrcode_image)
```

We can avoid the `kwarg` if it is only used once

maharshivpatel committed:
code optimisation

removed redundant ( unnecessary ) checks and kwarg

Co-authored-by: Saqib Ansari <nextchamp.saqib@gmail.com>

maharshivpatel committed:
fix linter and removed update_url

there aren't enough use cases so i have removed update_url from attach_qrcode_image function and some linter fix.

nextchamp-saqib commented:
> currently, there is no way to change anything in the invoice once you generate IRN however this is not the correct approach as there are many fields that have nothing to do with Invoice details that we send to the IRP ( government ) and we should be able to change fields that don't affect data we sent to the IRP. I am working on it and will send a separate PR or this.

I had implemented a [validation](https://github.com/frappe/erpnext_gst_compliance/blob/bd78ba4556f524fa4f8a6a87a2f21e6b202aa37c/erpnext_gst_compliance/erpnext_gst_compliance/doctype/e_invoice/e_invoice.py#L667-L691) for the same, it compares the e-invoice JSON before and after a field was modified, if the before & after JSON was the same, then the modification was allowed. 

However, this isn't needed anymore right? Once the e-invoice is generated on invoice submission, the invoice need not be edited anymore

Or do you foresee any other scenario where this validation would be needed?

maharshivpatel commented:
> Or do you foresee any other scenario where this validation would be needed?

Current `if irn` will make the **"allow on submit"** fields not editable so this [validation](https://github.com/frappe/erpnext_gst_compliance/blob/bd78ba4556f524fa4f8a6a87a2f21e6b202aa37c/erpnext_gst_compliance/erpnext_gst_compliance/doctype/e_invoice/e_invoice.py#L667-L691) will be needed.


For example, The issue that I am working on right now is that the user doesn't want to create the e-way bill as the invoice value is less than 50000.

if the user adds transport details then an e-way bill will be generated so the temporary fix would be the user would add it after IRN generation but this is not allowed due to the `if irn` statement.

>  I am working on a more user-friendly solution for the above issue. 


There is [e_inovice](https://github.com/frappe/erpnext/tree/develop/erpnext/regional/india/e_invoice) inside regional that is supposed to be removed in v14 then there is the **[erpnext_gst_compliance](https://github.com/frappe/erpnext_gst_compliance)** app and maybe there is a private repo that @deepeshgarg007  mentioned [here](https://github.com/frappe/erpnext/pull/30908#issuecomment-1119214422) so which one of these will be available in v14 as in erpnext_gst_compliance there aren't any commits after Jan 6, 2022, so it lacks some of the important fixes and is not tested yet in production. we (the public) don't have access to the repo that deepeshgarg007 mentioned

nextchamp-saqib commented:
> Current `if irn` will make the **"allow on submit"** fields not editable so this [validation](https://github.com/frappe/erpnext_gst_compliance/blob/bd78ba4556f524fa4f8a6a87a2f21e6b202aa37c/erpnext_gst_compliance/erpnext_gst_compliance/doctype/e_invoice/e_invoice.py#L667-L691) will be needed.

We can remove the validation altogether. If IRN will be generated on submission, then modifying an invoice value after submission is not possible. So is there a need to keep the validation that some fields of the invoice cannot be modified?

> so which one of these will be available in v14

Currently, the changes in the current `develop` will be available in v14. Both of those apps are not yet in production.
